WTG.Analyzers/Analyzers/PublicTupleType/PublicTupleTypeAnalyzer.cs

```diff
@@ -0,0 +1,303 @@
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using WTG.Analyzers.Utils;
+
+namespace WTG.Analyzers
+{
+	[DiagnosticAnalyzer(LanguageNames.CSharp)]
+	public sealed partial class PublicTupleTypeAnalyzer : DiagnosticAnalyzer
+	{
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(
+			Rules.AvoidTupleTypesInPublicInterfacesRule);
+
+		public override void Initialize(AnalysisContext context)
+		{
+			context.EnableConcurrentExecution();
+			context.RegisterCompilationStartAction(CompilationStart);
+		}
+
+		void CompilationStart(CompilationStartAnalysisContext context)
+		{
+			var cache = new FileDetailCache();
+
+			context.RegisterSyntaxNodeAction(
+				c => Analyze(c, cache),
+				SyntaxKind.ConstructorDeclaration,
+				SyntaxKind.MethodDeclaration,
+				SyntaxKind.PropertyDeclaration,
+				SyntaxKind.IndexerDeclaration,
+				SyntaxKind.DelegateDeclaration,
+				SyntaxKind.EventDeclaration,
+				SyntaxKind.EventFieldDeclaration,
+				SyntaxKind.FieldDeclaration);
+		}
+
+		static void Analyze(SyntaxNodeAnalysisContext context, FileDetailCache cache)
+		{
+			if (cache.IsGenerated(context.SemanticModel.SyntaxTree, context.CancellationToken))
+			{
+				return;
+			}
+
+			var decl = (MemberDeclarationSyntax)context.Node;
+			var processor = new Processor(context.SemanticModel, context.ReportDiagnostic, context.CancellationToken);
```

Comment:
A new allocation for each interrogated node? 😢

Reply:
Not every node, every externally visible member ...

Unfortunately there isn't a SyntaxNode visitor that takes an argument in the visit methods, which means the object needs state. Once the visitor has state, reusing them becomes problematic. I'll play around with it tonight to see if I can improve things by inverting a few responsibilities.

Reply:
I moved the checks out of the visitor so that we only need to create the visitor once we determine that the member is not allowed to use tuple types.